File: leetcode/python/606_construct_string_from_binary_tree.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def tree2str(self, root: TreeNode | None) -> str:
        def postorder(root):
            if not root:
                return ''
            if not root.left and not root.right:
                return str(root.val)
            if not root.right:
                return str(root.val) + '(' + postorder(root.left) + ')'
            return str(root.val) + '(' + postorder(root.left) + ')(' + postorder(root.right) + ')'

        return postorder(root)
        

File: leetcode/rust/src/606_construct_string_from_binary_tree.rs
// Definition for a binary tree node.
#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
  pub val: i32,
  pub left: Option<Rc<RefCell<TreeNode>>>,
  pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
  #[inline]
  pub fn new(val: i32) -> Self {
    TreeNode {
      val,
      left: None,
      right: None
    }
  }
}

use std::rc::Rc;
use std::cell::RefCell;

impl Solution {
    pub fn tree2str(root: Option<Rc<RefCell<TreeNode>>>) -> String {
        Solution::preorder(root)
    }

    fn preorder(root: Option<Rc<RefCell<TreeNode>>>) -> String {
        if let Some(node) = root {
            let node = node.borrow();
            let mut s = String::new();
            s.push_str(&node.val.to_string());
            if node.left.is_some() || node.right.is_some() {
                s.push('(');
                s.push_str(&Self::preorder(node.left.clone()));
                s.push(')');
            }
            if node.right.is_some() {
                s.push('(');
                s.push_str(&Self::preorder(node.right.clone()));
                s.push(')');
            }
            s
        } else {
            String::new()
        }
    }
}
